### ISSUE TITLE
Invalid flags supplied to RegExp constructor 'y' 

### ISSUE THREAD

**Issue by fernandogmar** [projectfluent/fluent.js#129]
Hi there!

I got this error on safari 9.1

It looks like you use this line on the parser.js
`const identifierRe = new RegExp('[a-zA-Z_][a-zA-Z0-9_-]*', 'y');`

the sticky flag is not supported...
I have seen this but I haven t tried yet
https://www.npmjs.com/package/babel-plugin-transform-es2015-sticky-regex

**Comment by fernandogmar** [comment]
I couldn't get it working... using babel transform and babel polyfills.... any idea/workaround?
Do you have plan to add backward compatibility with older browsers as safari v9?

**Comment by stasm** [comment]
Hi @fernandogmar, thanks for the report.

It looks like the [babel plugin only transforms](https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-sticky-regex) the regex literal syntax `/a/y` to a `new RegExp("a", "y")` constructor. It still relies on the browser's support for the sticky flag. The [support for the sticky flag is pretty good]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky#Browser_compatibility) these days. It's part of ECMAScript 2015 specification—it's already 3 years old!

I tried the patch below which passed the tests but regressed the performance of the runtime parser by +4%. @zbraniecki Do you have an opinion on this? According to _Can I use_ pre-10 versions of Safari have around 0.5% of the market, with Safari 9.1 at 0.29%.

I'd like `fluent` to have a similar compatibility story as React. For this reason I defined the browser list for the compat builds as:

https://github.com/projectfluent/fluent.js/blob/0b8ba5f7548d1ccd3d0dd79340380d66a9472adf/compat_config.js#L16-L21

This currently evaluates to http://browserl.ist/?q=%3E1%25%2C+last+4+versions%2C+Firefox+ESR%2C+not+ie+%3C+9 which includes Safari 9.1.

----

My test patch against `master`. Is there a better way to this without regressing performance?

```diff
diff --git a/fluent/src/parser.js b/fluent/src/parser.js
index e6387ce..1760240 100644
--- a/fluent/src/parser.js
+++ b/fluent/src/parser.js
@@ -1,14 +1,14 @@
 /*  eslint no-magic-numbers: [0]  */
 
 const MAX_PLACEABLES = 100;
 
-const privateIdentifierRe = new RegExp('-?[a-zA-Z][a-zA-Z0-9_-]*', 'y');
-const publicIdentifierRe = new RegExp('[a-zA-Z][a-zA-Z0-9_-]*', 'y');
+const privateIdentifierRe = new RegExp('-?[a-zA-Z][a-zA-Z0-9_-]*');
+const publicIdentifierRe = new RegExp('[a-zA-Z][a-zA-Z0-9_-]*');
 const functionIdentifierRe = /^[A-Z][A-Z_?-]*$/;
 
 /**
  * The `Parser` class is responsible for parsing FTL resources.
  *
  * It's only public method is `getResource(source)` which takes an FTL string
  * and returns a two element Array with an Object of entries generated from the
  * source as the first element and an array of SyntaxError objects as the
@@ -227,26 +227,26 @@ class RuntimeParser {
 
   /**
    * Get identifier of a Message, Attribute or External Attribute.
    *
    * @returns {String}
    * @private
    */
   getIdentifier(re) {
-    re.lastIndex = this._index;
-    const result = re.exec(this._source);
+    const result = re.exec(this._source.slice(this._index));
 
     if (result === null) {
       this._index += 1;
       throw this.error(`Expected an identifier [${re.toString()}]`);
     }
 
-    this._index = re.lastIndex;
-    return result[0];
+    const [id] = result;
+    this._index += id.length;
+    return id;
   }
 
   /**
    * Get a potentially private identifier (staring with a dash).
    *
    * @returns {String}
    * @private
    */
```

**Comment by Pike** [comment]
I don't think our compat_config is any useful, as mentioned before. It doesn't do what it looks like, to begin with.

@fernandogmar, what Safari 9.1 are you talking about, and why did you end up there? iOS or macos? particular hardware or OS version?

**Comment by fernandogmar** [comment]
It has been tested on browserstack Mac/El Capitan/Safari 9.1/
We were using l20n.js but we switched to fluent.js to move forward on our platform. As @stasm mentioned Safari have around 0.5% of the market, with Safari 9.1 at 0.29%... same goes for our current users, we are discussing what is best to do from here that is why I was asking...

**Comment by stasm** [comment]
> I don't think our compat_config is any useful, as mentioned before. It doesn't do what it looks like, to begin with.

I disagree. Perhaps it's a bit confusing how multiple queries are parsed (it seems that they are `OR`-ed unless they're negated in which case they are `AND`-ed with all previous queries) but the list if useful in understanding what browser we target and which transforms should we applied by Babel.

**Comment by stasm** [comment]
After a long discussion in #133 we settled on the following list of supported browsers for the `compat` builds:

    "Firefox >= 52",
    "FirefoxAndroid >= 52",
    "Chrome >= 55",
    "ChromeAndroid >= 55",
    "Edge >= 15",
    "Safari >= 10.1",
    "iOS >= 10.3",

These are all the browsers which natively support async functions. Safari 9 is not one of them. I realize that this decision will make it harder for some projects to use Fluent. Supporting these old browsers came with an engineering cost which was too significant for our small team. In particular, I got stung more than once by `regeneratorRuntime` being declared as a global in compat builds. Choosing to support only the browser on the list above solves this problem (and many other) completely: by removing the need to even use `regeneratorRuntime` in the first place.

The change landed in #225. It will go live in next releases of each package in the `fluent` family. I'm closing this issue: we officially don't support Safari 9 now.
